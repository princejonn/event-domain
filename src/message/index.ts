export * from "./Command";
export * from "./DomainEvent";
export * from "./Message";
export * from "./Timeout";
