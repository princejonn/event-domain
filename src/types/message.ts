import { AggregateIdentifier } from "./aggregate";
import { MessageType } from "../enum";

export interface IMessage {
  aggregate: AggregateIdentifier;
  causationId: string;
  correlationId: string;
  data: Record<string, any>;
  delay: number;
  id: string;
  mandatory: boolean;
  name: string;
  timestamp: Date;
  type: MessageType;
}

export interface MessageOptions {
  aggregate: AggregateIdentifier;
  causationId?: string;
  correlationId?: string;
  data?: Record<string, any>;
  delay?: number;
  id?: string;
  mandatory?: boolean;
  name: string;
  timestamp?: Date | string;
  type?: MessageType;
}
