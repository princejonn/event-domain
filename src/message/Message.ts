import { AggregateIdentifier, IMessage, MessageOptions } from "../types";
import { MessageType } from "../enum";
import { isString } from "lodash";
import { parseISO } from "date-fns";
import { randomUUID } from "crypto";

export abstract class Message implements IMessage {
  readonly id: string;
  readonly name: string;
  readonly type: MessageType;

  readonly aggregate: AggregateIdentifier;
  readonly causationId: string;
  readonly correlationId: string;
  readonly data: Record<string, any>;
  readonly delay: number;
  readonly mandatory: boolean;
  readonly timestamp: Date;

  protected constructor(options: MessageOptions, causation?: IMessage) {
    this.id = options.id || randomUUID();
    this.name = options.name;
    this.type = options.type || MessageType.UNKNOWN;

    this.aggregate = options.aggregate;
    this.causationId = options.causationId || causation?.id || this.id;
    this.correlationId = options.correlationId || causation?.correlationId || randomUUID();
    this.data = options.data || {};
    this.delay = options.delay || 0;
    this.mandatory = options.mandatory || false;

    const timestamp = options.timestamp || new Date();
    this.timestamp = isString(timestamp) ? parseISO(timestamp) : timestamp;
  }
}
