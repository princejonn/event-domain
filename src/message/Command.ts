import { MessageOptions, IMessage } from "../types";
import { Message } from "./Message";
import { MessageType } from "../enum";

export class Command extends Message implements IMessage {
  constructor(options: MessageOptions, causation?: Message) {
    super(
      {
        ...options,
        mandatory: true,
        type: MessageType.COMMAND,
      },
      causation,
    );
  }
}
