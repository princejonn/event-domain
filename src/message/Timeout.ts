import { IMessage, MessageOptions } from "../types";
import { Message } from "./Message";
import { MessageType } from "../enum";

export class Timeout extends Message implements IMessage {
  constructor(options: MessageOptions, causation?: Message) {
    super(
      {
        ...options,
        mandatory: true,
        type: MessageType.TIMEOUT,
      },
      causation,
    );
  }
}
