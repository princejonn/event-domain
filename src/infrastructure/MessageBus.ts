import { IMessageBus, MessageBusOptions, Subscription } from "../types";
import { Logger } from "@lindorm-io/winston";
import { Message } from "../message";
import { MessageBusInMemory } from "./in-memory";
import { MessageBusType } from "../enum";
import { PUBLISH_SCHEMA, SUBSCRIBE_SCHEMA } from "../schema";
import { validateSchemaAsync } from "../util";
import { MessageBusAMQP } from "./amqp";

export class MessageBus {
  private readonly logger: Logger;
  private messageBus: IMessageBus;

  constructor(logger: Logger, options?: MessageBusOptions) {
    this.logger = logger.createChildLogger(["MessageBus"]);

    const type = options?.type || MessageBusType.MEMORY;

    switch (type) {
      case MessageBusType.AMQP:
        this.messageBus = new MessageBusAMQP(this.logger, options.amqp, options.options);
        break;

      case MessageBusType.CUSTOM:
        this.messageBus = options.custom;
        break;

      case MessageBusType.MEMORY:
        this.messageBus = new MessageBusInMemory();
        break;

      default:
        throw new Error("Invalid message bus type");
    }
  }

  async publish(messages: Array<Message>): Promise<void> {
    this.logger.debug("publish messages", { messages });

    await validateSchemaAsync(PUBLISH_SCHEMA, { messages });

    await this.messageBus.publish(messages);
  }

  async subscribe(subscription: Subscription): Promise<void> {
    this.logger.debug("add subscription", { subscription });

    await validateSchemaAsync(SUBSCRIBE_SCHEMA, { subscription });

    await this.messageBus.subscribe(subscription);
  }
}
