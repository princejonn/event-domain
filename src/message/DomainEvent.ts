import { Message } from "./Message";
import { IMessage, MessageOptions } from "../types";
import { Command } from "./Command";
import { MessageType } from "../enum";

export class DomainEvent extends Message implements IMessage {
  constructor(options: MessageOptions, causation?: Command) {
    super(
      {
        ...options,
        type: MessageType.EVENT,
      },
      causation,
    );
  }
}
