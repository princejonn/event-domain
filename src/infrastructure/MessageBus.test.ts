import MockDate from "mockdate";
import { Command, DomainEvent } from "../message";
import { MessageBus } from "./MessageBus";
import { logger } from "../test";

MockDate.set("2022-01-01T08:00:00.000Z");

jest.mock("crypto", () => ({
  randomUUID: jest.fn(() => "e64b576a-022e-4dd4-8c97-32408012f8d1"),
}));

describe("MessageBus", () => {
  const aggregateIdentifier = {
    id: "2a553a74-0f84-4a7d-b91e-2902666db98c",
    name: "name",
    context: "context",
  };

  let messageBus: any;

  beforeEach(() => {
    messageBus = new MessageBus(logger);
  });

  describe("constructor", () => {
    test("should create a message bus", () => {
      expect(new MessageBus(logger)).toStrictEqual(expect.any(MessageBus));
    });

    test("should throw if message bus type is erroneous", () => {
      expect(
        () =>
          new MessageBus({
            // @ts-ignore
            type: "wrong",
          }),
      ).toThrow();
    });
  });

  describe("subscribe", () => {
    let callback: any;
    let handler: any;

    beforeEach(() => {
      callback = jest.fn();
      handler = jest.fn();
    });

    test("should add subscriber", async () => {
      await expect(
        messageBus.subscribe({
          callback,
          sub: {
            name: "event1",
            aggregate: {
              name: aggregateIdentifier.name,
              context: aggregateIdentifier.context,
            },
            handler,
          },
          topic: {
            name: "event1",
            aggregate: {
              name: aggregateIdentifier.name,
              context: aggregateIdentifier.context,
            },
            handler,
          },
        }),
      ).resolves.toBe(undefined);
      expect(messageBus.messageBus.subscriptions).toMatchSnapshot();
    });
  });

  describe("publish", () => {
    let callback: any;

    beforeEach(() => {
      callback = jest.fn().mockResolvedValue(undefined);
    });

    test("should evoke the subscription callback", async () => {
      const command1 = new Command({
        id: "86e9a7ee-7455-4438-9c61-7e41bd0191d5",
        name: "command1",
        aggregate: aggregateIdentifier,
      });
      const event1 = new DomainEvent(
        {
          id: "6d35d053-fc89-4ee5-93d6-87c0ceecb8b3",
          name: "event1",
          aggregate: aggregateIdentifier,
          data: { eventData: true },
        },
        command1,
      );
      await messageBus.subscribe({
        callback,
        sub: {
          name: "event1",
          aggregate: {
            name: aggregateIdentifier.name,
            context: aggregateIdentifier.context,
          },
        },
        topic: {
          name: "event1",
          aggregate: {
            name: aggregateIdentifier.name,
            context: aggregateIdentifier.context,
          },
        },
      });

      await expect(messageBus.publish([event1])).resolves.toBe(undefined);
      expect(callback).toHaveBeenCalledTimes(1);
    });
  });
});
