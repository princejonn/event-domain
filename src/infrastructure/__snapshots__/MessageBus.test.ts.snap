// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageBus subscribe should add subscriber 1`] = `
Array [
  Object {
    "callback": [MockFunction],
    "sub": Object {
      "aggregate": Object {
        "context": "context",
        "name": "name",
      },
      "handler": [MockFunction],
      "name": "event1",
    },
    "topic": Object {
      "aggregate": Object {
        "context": "context",
        "name": "name",
      },
      "handler": [MockFunction],
      "name": "event1",
    },
  },
]
`;
